doc/modules/mod_stream_management.md

```diff
@@ -1,21 +1,100 @@
+stream_mgmt_constraint_check_tref, %% another ref() for a timeout, this time for buffer_full check
+```
+
+### In `mod_stream_management`
+
+This module is just a "starter", to supply the configuration values to new client connections. It
+also provides a basic session table API and adds a new stream feature.
+
+At a bare minimum, this module keeps the config values in its `gen_mod` records, and it keeps a
+mnesia table defined as follows:
+
+```erlang
+-record(sm_session,
+        {smid :: smid(),
+         sid :: ejabberd_sm:sid()
+        }).
+```
+
+Where `smid` is a unique identifier — in this case a random binary, and `sid` is an opaque session
+identifier from `ejabberd_sm`, which is needed to find the previous session we want to resume from.
+This module implements hooks that run on connection removals and session cleanups, in order to clean
+records from a dying session; and it also implements registration callbacks, used in `ejabberd_c2s`
+when a session is registered for resumption.
+
+In order to be compliant with the XEP version 1.6, where a server can do the best effort to give the
+user the value of the server's `<h>` when a session timed out and cannot be resumed anymore, there's
+a second optional table:
+
+```erlang
+-record(stream_mgmt_stale_h,
+        {smid :: smid(),
+         h :: non_neg_integer(),
+         stamp :: non_neg_integer()
+        }).
+```
+
+This table is created, together with a `gen_server` responsible for cleaning up the tables, when
+`stale_h` is set to true with the proper garbage collection configuration. Then, when removing a
+record from the `sm_session` table (which happens when the state of the previous session is also
+dropped), a new record is added to this new table with the `smid` and `h` values of the dropped
+session, together with a timestamp. Then, when a new session attempting resumption queries
+`mod_stream_management` for the data behind a `smid`, `mod_stream_management` can answer one of the
+following:
+
+```erlang
+{sid, ejabberd_sm:sid()} | {stale_h, non_neg_integer()} | {error, smid_not_found}.
+```
+
+And `ejabberd_c2s` will pattern-match and act accordingly.
 
 ### Options
 
 * `buffer_max` (default: 100): Buffer size for messages yet to be acknowledged.
-* `ack_freq` (default: 1): Frequency of ack requests sent from the server to the client, e.g. 1 means a request after each stanza, 3 means a request after each 3 stanzas.
-* `resume_timeout` (default: 600): Timeout for the session resumption. Sessions will be removed after the specified number of seconds.
+* `ack_freq` (default: 1): Frequency of ack requests sent from the server to the client, e.g. 1
+  means a request after each stanza, 3 means a request after each 3 stanzas.
+* `resume_timeout` (default: 600): Timeout for the session resumption. Sessions will be removed
+  after the specified number of seconds.
+* `stale_h`: enable keeping old server's `<h>` values after the resumption timed out. Defaults to
```

Comment:
We try to avoid non-tagged options (i.e. don't use tuples for options, because it would make harder implementing better config).

Good alternatives for you to choose from:
- use flat options: i.e. stale_h_enabled, stale_h_gc_repeat_after, stale_h_gc_geriatric.
- use {stale_h_opts, [{enabled, true}, {stale_h_gc_repeat_after, ...}, ...}]}
- use stale_h_opts and stale_h_enabled
- assume if stale_h_opts are defined, stale_h is enabled.